### PR TITLE
Fix manifest precedence

### PR DESCRIPTION
**Pull request by asmello**
The manifest file takes precedence over the lockfile when specifying where to download an artefact.

### CODE REVIEW

**Comment by asmello** [comment]
Fixes #76

File: src/command.rs [hunk 1]
```diff
@@ -93,9 +93,13 @@ pub async fn add(registry: RegistryUri, dependency: &str) -> eyre::Result<()> {
...
 
-    manifest
-        .dependencies
-        .push(Dependency::new(registry, repository, package, version));
+    manifest.dependencies.push(Dependency::new(
+        registry,
+        repository,
+        package,
+        version,
+        crate::manifest::DependencyType::Root,
```
Comment (mara-schulke) [inline src/command.rs:101]: root is not serialized i guess?
why is it needed?

Comment (asmello) [reply src/command.rs:101]: It's not needed here, it's used to build the dependency graph while allowing manifest-derived dependencies to override the lockfile metadata, but I haven't thought about serialisation so I'll confirm if that works sensibly.

**Comment by mara-schulke** [comment]
I feel like we can undo the "dependencytype" and still solve the manifest precedence?

File: src/command.rs (hunk 1 above)
Comment (asmello) [reply src/command.rs:101]: Ok, just checked and yes it gets ignored as it's supposed to.

**Comment by asmello** [comment]
> I feel like we can undo the "dependencytype" and still solve the manifest precedence?

Hmm, that was my initial idea but it felt awkward to carry extra dependency metadata around separate of the dependency struct, but since this is only really relevant in one place I think that might be best after all. Will look into how that looks like in practice.

File: src/command.rs (hunk 1 above)
Comment (asmello) [reply src/command.rs]: Removed it after all. 🤷

File: src/manifest.rs [hunk 2]
```diff
@@ -178,6 +178,20 @@ impl Dependency {
             },
         }
     }
+
+    pub fn with_version(&self, version: &Version) -> Dependency {
```
Comment (mara-schulke) [inline src/manifest.rs:182]: Same here

File: src/package.rs [hunk 3]
```diff
@@ -584,15 +592,23 @@ impl DependencyGraph {
...
+            let package = registry
+                .download(dependency.with_version(&local_locked.version))
+                .await
+                .wrap_err_with(|| {
+                    format!(
+                        "Error downloading package {}@{}",
+                        local_locked.name, local_locked.version
+                    )
+                })?;
             local_locked.validate(&package).wrap_err_with(|| {
```
Comment (mara-schulke) [inline src/package.rs:612]:
```suggestion

            local_locked.validate(&package).wrap_err_with(|| {
```

**Review by mara-schulke — APPROVED** [review]
Approved, but when reading through the algorithm it seems very complex! Maybe we could chat the next days on how to simplify it eventually

**Comment by asmello** [comment]
> Approved, but when reading through the algorithm (not written in this pr tho!) it seems very complex! Maybe we could chat the next days on how to simplify it eventually

I'm skeptical it can be made any simpler (I tried hard to make it as simple as possible for the V0). I actually have ideas for a slightly more complex but much more intelligent version using a trick similar to what Poetry does. But let's chat, I'm not planning on pursuing any optimisations in the immediate future but we should align on vision.

**Pull request merged**